python_modules/dagster/dagster/_core/definitions/asset_graph.py

```diff
@@ -310,18 +313,32 @@ def get_parent_partition_keys_for_child(
 
         return parent_partitions_def.get_partition_keys_in_range(upstream_partition_key_range)
 
-    def get_roots(self, asset_key: AssetKey) -> AbstractSet[AssetKey]:
-        """Returns all root assets that the given asset depends on"""
-        if not self.get_parents(asset_key):
+    def has_non_source_parents(self, asset_key: AssetKey) -> bool:
+        """Determines if an asset has any parents which are not source assets"""
+        if asset_key in self.source_asset_keys:
+            return False
+        return bool(self.get_parents(asset_key) - self.source_asset_keys)
+
+    def get_non_source_roots(self, asset_key: AssetKey) -> AbstractSet[AssetKey]:
+        """Returns all assets upstream of the given asset which do not consume any other
+        AssetsDefinitions (but may consume SourceAssets).
+        """
+        if not self.has_non_source_parents(asset_key):
             return {asset_key}
-        return {key for key in self.upstream_key_iterator(asset_key) if not self.get_parents(key)}
+        return {
+            key
+            for key in self.upstream_key_iterator(asset_key)
+            if not self.has_non_source_parents(key)
+        }
 
     def upstream_key_iterator(self, asset_key: AssetKey) -> Iterator[AssetKey]:
         """Iterates through all asset keys which are upstream of the given key."""
         visited: Set[AssetKey] = set()
         queue = deque([asset_key])
         while queue:
             current_key = queue.popleft()
+            if current_key in self.source_asset_keys:
```

Comment:
Why is this needed? Won't the call to `self.get_parents` on the next line return an empty iterable?

Reply:
the dependency dictionary doesn't contain a key for source assets, and get_parents doesn't use a get() call (it just accesses the dictionary directly). arguably, it should use .get(), was just trying not to disturb existing logic too much